### PR TITLE
[BD-21] Extract waffle_utils from edx-platform

### PR DESCRIPTION
**Pull request by regisb**
**Description:** Extract the various waffle switch/flag classes from edx-platform such that they can be used in other IDAs. This implements ADR `#2 <https://github.com/edx/edx-toggles/tree/master/docs/decisions/0002-application-toggle-state.rst>`__, `#3 <https://github.com/edx/edx-toggles/tree/master/docs/decisions/0003-django-setting-toggles.rst>`__, `#4 <https://github.com/edx/edx-toggles/tree/master/docs/decisions/0004-toggle-api.rst>`__. These changes are backward-compatible. Suppression of the namespace classes will happen in a separate PR.

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943?src=mail&src.mail.timestamp=1593719976942

**Reviewers:**
- [ ] @robrap 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

This is a work in progress. A couple things still need to be done:
- [x] Move waffle flag classes
- [x] Move unit tests
- [ ] Document the new classes
- [x] Do not export BaseNamespace, which should not be exposed to edx-platform

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-617](https://openedx.atlassian.net/browse/BLENDED-617) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

File: edx_toggles/toggles/internal/__init__.py
Comment (robrap) [inline edx_toggles/toggles/internal/__init__.py:3]: Could you move this to a README.rst instead?

File: edx_toggles/toggles/internal/base.py
Comment (robrap) [inline edx_toggles/toggles/internal/base.py:12]:
```suggestion
    ``_class_instances`` class attribute, which is exposed via the ``get_instances`` class method.
```

File: edx_toggles/toggles/internal/base.py
Comment (robrap) [inline edx_toggles/toggles/internal/base.py:34]: Please see https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/waffle_utils/docs/decisions/0004-waffle-util-namespacing.rst  There are 3 ADRs relevant to the extraction.  I'm sorry if I didn't make that clear.

**Review by robrap — COMMENTED** [review]
I'm going to wait on further review.

File: edx_toggles/toggles/internal/waffles.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffles.py:2]: Nit: I'd prefer `waffle.py` for the name of the file.

File: edx_toggles/toggles/internal/waffles.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffles.py:83]: I'm hoping this can go away, but needs some investigation.  I feel like this was added at the beginning, and used with initial grades rollout (if I'm remembering correctly), but I don't think this is used in general and I'm not clear why all of this is needed.  I don't imagine most switches use this.

File: edx_toggles/toggles/internal/base.py
Comment (robrap) [reply edx_toggles/toggles/internal/base.py:34]: @nasthagiri: FYI: I am planning on the waffle namespace classes going away.  Please see the ADR to understand why.  Since you added namespacing to OEP-17, and were involved in the initial design (I think), it would probably be good for you to read and agree to the [ADR above dropping the namespacing classes](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/waffle_utils/docs/decisions/0004-waffle-util-namespacing.rst).

**Review by robrap — COMMENTED** [review]
I added a couple of additional high-level comments.  Thanks!

File: edx_toggles/toggles/internal/base.py
Comment (nasthagiri) [reply edx_toggles/toggles/internal/base.py:34]: @robrap Thanks for the heads up!  I'm more concerned about ensuring developers namespace their toggles, than the mechanism by which we enforce this. Looks like the ADR covers this via an assertion of the existence of a `.` in the name.

File: edx_toggles/toggles/internal/__init__.py
Comment (regisb) [reply edx_toggles/toggles/internal/__init__.py:3]: Are you sure? This comment will only be read by developers, so it would make sense to keep it as near as possible to the code (IMHO).

File: edx_toggles/toggles/internal/waffles.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffles.py:2]: :ok_hand:

File: edx_toggles/toggles/internal/waffles.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffles.py:83]: Ok I'm adding this to the list of things to be investigated later.

**Comment by regisb** [comment]
@robrap I expect that the suppression of the namespace classes is going to be tedious, but not otherwise complex. So I propose that we resolve the extraction of the waffle and switch classes first. I will then open a new PR to get rid of the many namespace instances. Otherwise we will end up with a PR that touches upon dozens of files, which is inconvenient to review.

**Comment by regisb** [comment]
@robrap This should now be ready to review.

File: edx_toggles/toggles/internal/__init__.py
Comment (robrap) [reply edx_toggles/toggles/internal/__init__.py:3]: **tl;dr:** You can leave this as-is.

**Longer:**
I'm not sure. :) I used to hate `__init__.py` files because you never knew if they were empty or not, so you never knew whether to look in them.  Now that we are on a newer version of Python, can we remove them if they are empty?  Assuming so, I have less of an issue with docs being in this file. :)

File: edx_toggles/toggles/internal/waffles.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffles.py]: FYI: This was moved back to edx-platform.

File: edx_toggles/toggles/internal/waffle.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle.py:164]: Maybe?:
```
Return and cache the value of the flag activation without adding monitoring.
```

File: edx_toggles/toggles/internal/waffle.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle.py:146]: Maybe?:
```
Returns, caches, and adds monitoring for whether the provided flag is active.
```

File: edx_toggles/toggles/internal/waffle.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle.py:192]: This name and docstring don't really match the implementation, which includes lots of warnings, because this is an exceptional case.  What about `_get_flag_active_for_missing_request` or `_get_flag_active_default_for_missing_request`?

File: edx_toggles/toggles/internal/waffle.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle.py:126]: Why were we able to lose this for WaffleFlag, but not WaffleSwitch?  Is that just a miss, or was this intentionally left here?

File: edx_toggles/urls.py
Comment (robrap) [inline edx_toggles/urls.py:6]: I don't actually know.  Is it ok to just delete this file?

File: CHANGELOG.rst
Comment (robrap) [reply CHANGELOG.rst:18]: We could also warn against using WaffleFlag and WaffleSwitch outside of edx-platform due to upcoming backward compatible changes,  but no one probably will.

File: CHANGELOG.rst
Comment (nedbat) [inline CHANGELOG.rst:17]: Is this fixing a consumer-visible problem? If so, we should mention which one in the changelog.  If not, it doesn't need to be in the changelog.

Comment (robrap) [reply CHANGELOG.rst:17]: @nedbat: FYI: @regisb may clarify this, but he's also working on the first PR to ever use this class, so it really won't affect anyone but him.

File: CHANGELOG.rst
Comment (robrap) [reply CHANGELOG.rst:18]: @regisb: I'll leave this to you.  I'm not going to block the PR on this.

File: edx_toggles/toggles/internal/__init__.py
Comment (robrap) [reply edx_toggles/toggles/internal/__init__.py:3]: I am marking this as resolved so I don't have to keep seeing it.

File: edx_toggles/toggles/internal/base.py
Comment (robrap) [reply edx_toggles/toggles/internal/base.py:56]: Again, I'm marking as resolved.  Hopefully you read this, but if not, I'll bring it up again in the future.

File: CHANGELOG.rst
Comment (nedbat) [reply CHANGELOG.rst:17]: Sounds like we don't need it in the changelog.

File: CHANGELOG.rst
Comment (regisb) [reply CHANGELOG.rst:18]: :ok_hand:

File: edx_toggles/toggles/internal/waffle.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle.py:224]: Are we prefixing all non-exposed function names by `_`? I think this adds a lot of cruft to the code, and in my understanding this is not the convention generally adopted for Open edX code. Anyway, I can modify the name to make it clearer that this function is not meant to be imported.

File: edx_toggles/toggles/internal/waffle.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle.py:339]: Yes, I understand. I tried to make it easier by moving all refactoring in a single commit in the edx-platform PR: https://github.com/edx/edx-platform/pull/25286/commits/d947456e7f3c84ccf34b79e51c60de75ff3af308

File: edx_toggles/toggles/internal/waffle.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle.py:306]: > so we leave that to another day (or lifetime).

:sweat_smile:

File: edx_toggles/toggles/internal/waffle.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle.py:312]: Indeed!

File: edx_toggles/toggles/internal/waffle.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle.py:192]: I'll go with `_get_flag_active_no_request` to be consisdent with the naming of the `_get_flag_active_request` method.

File: edx_toggles/toggles/internal/waffle.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle.py:126]: This is an oversight.

File: edx_toggles/urls.py
Comment (regisb) [reply edx_toggles/urls.py:6]: I'm not sure. I'd like to keep it to make it easier for other people to add urls there.

File: CHANGELOG.rst
Comment (regisb) [reply CHANGELOG.rst:17]: I think we do need it in the changelog, as edx-toggles==0.3.0 was already released on pypi. I'll improve the phrasing.

**Review by robrap — APPROVED** [review]
We will follow-up with cleaning out namespacing in a following PR.

**Pull request merged**
